### ISSUE TITLE
GO breaks sound after high playing

### ISSUE THREAD

**Issue by oleg68** [GrandOrgue/grandorgue#702]
This issue exists for a lot of time since very old versions.

When playing GO with lots of stops and notes and  high reveberation to a jack output (Rt:jack or Jack:native) sometimes sound stops. Trying to do Panic or to exit GO causes that GO hangs.

**Comment by oleg68** [comment]
After adding a diagnostic warning messages in #762, when the sound is brocken, there warning messages appeared in the log window:

```
Вс 17 окт 2021 18:23:51: 18:23:51: Warning: GOMutex: timeout when locking a mutex; currentLocker= newLocker=GOSoundGroupWorkItem::Run.beforeProcess
Вс 17 окт 2021 18:23:51: 18:23:51: Warning: GOMutex: timeout when locking a mutex; currentLocker= newLocker=GOSoundGroupWorkItem::Finish
Вс 17 окт 2021 18:23:51: 18:23:51: Warning: GOMutex: timeout when locking a mutex; currentLocker= newLocker=GOSoundGroupWorkItem::Run.beforeProcess
Вс 17 окт 2021 18:23:51: 18:23:51: Warning: GOMutex: timeout when locking a mutex; currentLocker= newLocker=GOSoundGroupWorkItem::Finish
```

The mutex could not be obtained in 1 second, but there were no any clear reason: there was no current locker. Moreover, the problem has been changed: earlier the sound stopped totally, now the sound returns after one second. It means that the mutex is obtained successfuly at the second attempt.

Both these facts suggest that the root cause of this problem is wrong implementation of GOMutex. Sometimes it cann't be locked when it is not locked by any other thread.

Now there are two alternative implementation of GOMutex and GOCondition in the GO source code: with the wxWirgets classes wxMutex, wxCondition and with GOWaitQueue. The implementations are switched by `WX_MUTEX` compilation variable.

GOWaitQueue is not a standard primitive and itself has two alternative implementations: with wxSemaphore and with std::mutex (has been switched to std::timed_mutex as part of #784).

Now by default GOMutex uses GOWaitQueue that uses wxSemaphore.

This code is too complex and it is hard to troubleshoot. Most likely it has a bug, that causes this issue.

When GO was created, there were no any standard primitives std::mutex and std::condition_variable. But [they appeared in `C++11`](https://en.cppreference.com/w/cpp/thread).

So seems the best solution is to **rewrite the current implementation of GOMutex and GOCondition with `std::timed_mutex` and `std::condition_variable`**. Very likely it will solve this issue.

**Comment by larspalo** [comment]
> Do you remember, had he tried the standard C++11 implementation?

I would need to dig deeper into the conversations on the developer list to possibly find an answer - or try asking him. I can say that the GOMutex comes around 2012 and that C++11 was not "enforced" until 2016 when some of the locking code did change but not everything obviously.

**Comment by oleg68** [comment]
Fixed with #837
